Core reverse implementation

thomasyip opened the pull request.

thomasyip committed:
Added nunjucks as lib.

thomasyip committed:
Used non-relative path for within storeys require().

thomasyip committed:
Updated jquery dep.

thomasyip committed:
Replaced dom if #id matches.

denti committed:
`get_url_patterns` was added
`reverse` blank was added

denti committed:
- storeys/utils/datastructures.js was added
- MultiValueDict was implemented

denti committed:
reverse() was implemented, but some dirty hacks should be fixed

thomasyip commented on storeys/js/core/urlresolver.js:
```diff
@@ -66,10 +66,99 @@ define(
       // -------------------------------------------
-      function reverse(viewname, params) {
+
+      function reverse(viewname, params, cb) {
```
It should simply return, instead of callback.

thomasyip commented on storeys/js/utils/datastructures.js:
```diff
@@ -0,0 +1,77 @@
+         *which returns a list for every key, even though most Web forms submit
+         *single name-value pairs.
+         */
+        function MultiValueDict(obj) {
```
Don't reimplement data structure. Just use JavaScript's hash.

denti replied:
I didn't found structure like `hash_tables` inside javascript (simple {} - is unsuitable because it doesn't supports multiple values for same key).  That's why I realized simple structure for work easy with hash tables.

thomasyip commented on storeys/js/core/urlresolver.js:
```diff
@@ -1,6 +1,6 @@
 define(
-    ['require', 'module', '../conf/urls'],
-    function(require, module, urls) {
+    ['require', 'jquery', 'settings', 'module', 'storeys/conf/urls'],
```
comment deleted. jquery was in my code. disregard.

denti committed:
- escape special characters for JS RegExp

denti committed:
- regexp fixs
- minor fixs

denti committed:
- encodeRFC3986URIComponent() was added
- Correct handling of `\`
- url for included patterns was changed in `reverse()`

denti committed:
- storeys/utils/datastructures was removed

denti committed:
- '`url` templatetag was added'

denti committed:
- lilfix for `url` templatetag

denti committed:
- lilfixes to error messages

denti committed:
IMPORTANT: problem of named parameters in RegExp should be fixed

- `urlresolver.resolve` was updated
- `urlresolver.reverse` was refactored
-  test for `resolve` were added

denti committed:
- submodule `xregexp` was  added `https://github.com/slevithan/xregexp`
- python regex named args parcer was added (using xregexp)
- tests were updated

thomasyip committed:
Updated nucjucks git path.

thomasyip committed:
Ported to use storeys fork of xregexp.
